### DISCUSSION TITLE
The Tldraw emmebed is not work

### DISCUSSION THREAD

**Discussion by Arstman** [outline/outline#4174] [General]
Looks like the Tldraw Integration not work?

I created a multiplayer project and paste the link as told, but it only shows this:
![图片](https://user-images.githubusercontent.com/14262700/192307179-c9e23b9f-b46e-4623-98a7-ff524817eb0f.png)

I have try both cloud and self-host, both same problem.

anything i have done wrong?

**Comment by Limezy** [comment]
Bad news !
I was the one doing the PR initially. Let me check tomorrow if there was some kind of breaking change on outline or tldraw side.

**Comment by Limezy** [comment]
I can reproduce the bug, and I could also reproduce it with another software (Obsidian).
It seems like a breaking change on tldraw side... Not allowing iframes anymore.
I've just asked on the official discord whether that is their intention or not... I really hope not !

**Reply by Arstman** [reply]
it seem not a good news...  :disappointed:

**Reply by Limezy** [reply]
<img width="726" alt="image" src="https://user-images.githubusercontent.com/24638389/192463169-83f174d1-34ed-476f-82a0-973d76874fa9.png">

I got a reply. While waiting for the proposed solutions - if they ever come... I can only think of two solutions to have Tldraw in Outline : 
* One would be to use Tldraw as it is originally supposed to work, that is as a nodejs plugin to be embedded inside Outline's codebase (a little bit like Mermaid that was recently added). That would be very nice, but it's *very much* out of my technical reach (I'm not a dev, very far from it actually!) and I'm not very sure @tommoor would accept it anyway, as it would raise up the overall Outline bundle size and memory footprint
* One would be to be able to self-host a multiplayer Tldraw instance, and have the Tldraw integration including an url parameter to be set in the settings, just like the draw.io self-hosted recent integration. I think this could be inside my technical capabilities, but I'm still not very clear whether it's possible to self-host a multiplayer Tldraw room or not. Now investigating further on that path

If any crazy dev reading us here would like to step in and try the first option, should Tom accept it that would be very nice !

**Reply by almereyda** [reply]
<details>

<summary>Optional comment</summary>

Due to the way how they implemented it

https://github.com/tldraw/tldraw/pull/800/files#diff-74ab2df42d7dcbc75e3e089e3e47c8e0af183d074c3189f1c7849fafa015bec4R17-R20

a simple CORS proxy would probably not help in this case. Maybe they can be asked to reimplement it in a more standard compliant way, in so they could block requests from known proxies, and allow others under some kind of fair-use agreement?

If they would have used [`X-Frame-Options`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options) instead, one could try to employ one of these:

- https://nordicapis.com/10-free-to-use-cors-proxies/
- https://gist.github.com/jimmywarting/ac1be6ea0297c16c477e17f8fbe51347
- https://www.ryadel.com/en/corsflare-free-cors-reverse-proxy-bypass-same-origin/

Or those:

* https://caddyserver.com/docs/caddyfile/directives/header
* https://doc.traefik.io/traefik/middlewares/http/headers/
* https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/
* https://gridpane.com/kb/how-to-add-your-own-custom-nginx-headers/
* https://serverfault.com/questions/678742/nginx-proxy-pass-rewrite-of-response-header-location
* https://github.com/http-party/node-http-proxy/issues/1401
* https://stackoverflow.com/questions/12755865/rewrite-response-headers-with-node-http-proxy
* https://github.com/10PinesLabs/CORS-X-frame-option-proxy
* 
</details>

**Reply by Limezy** [reply]
Good catch for the PR, thanks @almereyda 
I tend to understand the statement though "are footing the bill for someone else's product", so even if done nicely and communicating with them, your way seem to be a little bit hackish.
I would be more keen on "monkey patching" the [/apps/www](https://github.yanqishui.work/tldraw/tldraw/tree/main/apps/www) webapp that they provide, to self host it and have it accepting iFrames.
Would you have some hints on how to do that ? Would removing these lines (link removed) be enough ?

Edit : removed the bad link. I don't have any clue on how I landed in that fake GitHub fork...

**Reply by tommoor** [reply]
Yea, I'm not sure integrating TLDraw into the codebase is the way to go here. Would be happy to support self-hosted in the same way as Draw.io though. It would be lovely if we could work out an exception with TLDraw to continue to allow embedding but by the nature of these things that would probably only ever apply to the cloud hosted version if it was possible at all.

**Reply by Limezy** [reply]
@tommoor thanks for your reply. Yes, that makes sense for the fact that any agreement found between both products would only apply on the SaaS version.

I will try to self host tldraw first, and from there I may submit a PR to copy Draw.io integration style, with a custom URL.
I really want to have TLDraw back, as it was matching Outline UX so well !

Integrating TLDraw into the codebase would be great but I understand it may bring more problems than real usecases.

**Reply by almereyda** [reply]
While a fork always brings the additional burden to rebase one's fixes on their master, indeed that could work for a while.

The lines to remove would be

https://github.com/tldraw/tldraw/blob/d721ae6a2f72cbfbeaca9e75ab833f89a0ef19f7/apps/www/pages/r/%5Bid%5D.tsx#L19-L22

(Your link, @Limezy, pointed at a weird GitHub mirror site. I suggest to manually put `github.com` as domain, after one accidentally landed on such a page via a search engine.)

But I strongly suggest to seek contact with upstream, and find ways how this can be configured with a feature flag, in so they can simple enable the blocking, and others can manage it more liberally. Then, secondly, I would still favour a more standards-complaint way of handling this, such as with the `X-Frame-Options` header, and have their client-side JavaScript only be triggered, when that fails. But that remains open to discuss upstream. Thirdly, having the feature disabled on a per host basis, via an allow list, can also help to implement embedding for origins considered secure, e.g. within the same domain, while not opening up the room for unvoluntary reuse.

**Reply by Limezy** [reply]
Yep, thanks for your advice. 
I don't know what Tom will want to do with that problem, and I don't think I'm introduced enough to discuss anything serious with upstream.
However I've managed to self host my own tldraw with multiplayer using free api keys from liveblocks (unlimited rooms, max 5k connexions per room, 10 simultaneous editors, which is much more than I need). 
So I think I'll :
1. Propose a PR here with "draw.io style" self hosted integration
2. Build a Yunohost package for anybody needing a one click installation setup of a self hosted multiplayer tldraw

**Reply by Limezy** [reply]
The step 2 is now done : https://github.com/YunoHost-Apps/tldraw_ynh
Using Yunohost you can now self-host in one click a patched instance of tldraw so that you can use it with your own Liveblocks (free or not) API keys **and** use it within an iFrame without the new warning.
Now I will have to complete the step 1 and propose a PR so that one can use a self-hosted TLDraw instance instead of the tldraw.com one... Better than nothing !

**Reply by tommoor** [reply]
We need to find a way to combine all of these self-hosted integrations onto a single settings page, it doesn't make sense to keep adding items to the settings sidebar for these tiny apps that 99% of customers won't use.

**Reply by Limezy** [reply]
Yes, agreed !

**Reply by matbgn** [reply]
IMHO also keeping tldraw in the drop menu of possible integration is frustrating and contra productive with actual statement. 

And as far as I searched for quite a long time now, tldraw is the only compatible drawing app (excalidraw and okso are partially or totally based on tldraw) with Outline's design and UX philosophy. 

So my two cents here, would it not just be definitely better to have tldraw integrated natively like Mermaid? The use case of sketches in a wiki to explain complex situations is not only a corner case, as the discussions and issues, on the repo, demonstrated it.

**Reply by Limezy** [reply]
I would love to have tldraw integrated, and as you said the design and UX seem to be perfectly matching !
What I don't know is how hard it would be to use the current Outline's multiplayer backend with TLDraw.
It's clearly out of my reach as I'm not a programmer (very far from it actually !)

**Reply by thmsfchtnr** [reply]
That would be awesome! I love the idea, please developers make it possible 🙏🏼

**Comment by tommoor** [comment]
In the meantime it should be removed from the integrations menu - that one
didn’t last long

**Comment by Eweol** [comment]
Hi,

Any progress here?

**Reply by Vranda-Kansal** [reply]
any update here. Not able to integrate tldraw in iframe it shows visit this page on tldraw.com. Any Help?

**Reply by tommoor** [reply]
TLDraw has blocked embedding temporarily, there is nothing we can do about it but I encourage you to contact them. More details here:

https://github.com/outline/outline/issues/6706
